### ISSUE TITLE
SpatialLPPooling not working with norm 2

### ISSUE THREAD

**Issue by culurciello** [torch/DEPRECEATED-torch7-distro#62]
I do not think the SpatialLPPooling is working correctly:
see examples below
ok in case of 1-norm, not ok in case of 2-norm
Any comments:

---

t7> model = nn.Sequential()
t7> model:add(nn.SpatialLPPooling(1,1,2,2))
t7> inp = torch.zeros(1,7,7)
t7> out = model:forward(inp)
t7> =out
(1,.,.) = 
 0 0 0
 0 0 0
 0 0 0
[torch.DoubleTensor of dimension 1x3x3]

t7> model = nn.Sequential()
t7> model:add(nn.SpatialLPPooling(1,2,2,2))
t7> inp = torch.zeros(1,7,7)
t7> out = model:forward(inp)
t7> =out
(1,.,.) = 
 0.0001 *
   3.1623  3.1623  3.1623
   3.1623  3.1623  3.1623
   3.1623  3.1623  3.1623
[torch.DoubleTensor of dimension 1x3x3]

**Comment by clementfarabet** [comment]
It's not a bug really, it's just that there's an Epsilon that was added to the input of the Sqrt(), to avoid issues around the 0 (for the backprop). Not sure it it's really useful, from memory Koray added this?

**Comment by culurciello** [comment]
ok yes it seems to work for other values, but when I use it in a networks gives peculiar results

**Comment by koraykv** [comment]
I don't remember adding this to Torch, but I definitely have it in my kex module :). The main reason is at zero, the derivative is not defined due to division by zero. So I add a small factor (1e-7). If it causes any problems, it can be set to zero by

m:get(3).eps = 0

I haven't done detailed experiments to see if this would cause any problems, so if you are seeing problems, we can set it to zero by default and rely on people to set it to non-zero values to avoid getting NaNs.

**Comment by clementfarabet** [comment]
I don't really see how it can be an issue though. What would be the
alternative, writing a conditional gradient function?

**Comment by koraykv** [comment]
Well, I am not sure why it would be an issue. 1e-7 is such a small perturbation I don't think it would cause any problems in nn training. But, maybe it could for some cases...
Other implementations might be
1) Put an if in the backward (would slow down a lot)
2) Put a bias only in the gradient computation, the derivatives would be perturbed only
3) keep the current implementation, but make the default bias to zero (if you get NaNs in your model, this would be first place to look at

any other options?
